### PR TITLE
[4.0] Implement cleanup defunct silo entries ado net clustering

### LINKED ISSUE #7171: ADO.NET Clustering: implement `CleanupDefunctSiloEntries`
With Microsoft.Orleans.Clustering.AdoNet v3.4.3, and MySql.Data,  Microsoft Orleans handled SIGTERM signal and exited gracefully.
But the corresponding record in OrleansMembershipTable is not removed. Its status changes to `6`.

```
info: Orleans.Hosting.SiloHostedService[0]
      Stopping Orleans Silo
info: Orleans.Runtime.SiloLifecycleSubject[100452]
      OrleansDashboard.Metrics.GrainProfiler stopped in stage Last (2147483647) in 0.4551 Milliseconds.
info: Orleans.Runtime.SiloLifecycleSubject[100452]
      Stopping lifecycle stage Last (2147483647) took 2.9755 Milliseconds
info: Orleans.Runtime.SiloLifecycleSubject[100452]
      Orleans.Hosting.SiloBuilderStartupExtensions+StartupTask stopped in stage Active (20000) in 0.1449 Milliseconds.
info: Orleans.Runtime.SiloLifecycleSubject[100452]
      MembershipTableCleanupAgent stopped in stage Active (20000) in 1.6222 Milliseconds.
info: Orleans.Runtime.SiloLifecycleSubject[100452]
      ClusterHealthMonitor stopped in stage Active (20000) in 20.5276 Milliseconds.
info: Orleans.Runtime.SiloLifecycleSubject[100452]
      Orleans.Runtime.MembershipService.LocalSiloHealthMonitor stopped in stage Active (20000) in 10.6568 Milliseconds.
info: Orleans.Runtime.SiloLifecycleSubject[100452]
      Orleans.Runtime.Silo stopped in stage Active (20000) in 6.772 Milliseconds.
```

image

And I see the following warning message.
```
warn: Orleans.Runtime.MembershipService.MembershipTableCleanupAgent[100662]
      CleanupDefunctSiloEntries operation is not supported by the current implementation of IMembershipTable. Disabling the timer now.
```

The OrleansMembershipTable is bloating.  Is there a way to remove unneeded records from the table?

### PR DESCRIPTION
**Pull request by ElanHasson**
Resolves #7171.

Separate PR for 3.x. is #7929.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/orleans/pull/7928)

### CODE REVIEW

File: test/Extensions/TesterAdoNet/RelationalUtilities/RelationalStorageForTesting.cs [hunk 1]
```diff
@@ -114,6 +127,8 @@ public static async Task<RelationalStorageForTesting> SetupInstance(string invar
             // Concatenate scripts
             foreach (var fileName in testStorage.SetupSqlScriptFileNames)
             {
+                Console.WriteLine($"Processing {fileName}");
```
Comment (ReubenBond) [inline test/Extensions/TesterAdoNet/RelationalUtilities/RelationalStorageForTesting.cs:130]: We should either delete this or make it use `ITestOutputHelper` (an xUnit concept for logging during tests)

File: test/Extensions/TesterAdoNet/RelationalUtilities/PostgreSqlStorageForTesting.cs [hunk 2]
```diff
@@ -8,6 +8,8 @@ namespace Tester.RelationalUtilities
 {
     class PostgreSqlStorageForTesting : RelationalStorageForTesting
     {
+        protected override string ProviderMoniker => "PostgreSQL";
```
Comment (ReubenBond) [inline test/Extensions/TesterAdoNet/RelationalUtilities/PostgreSqlStorageForTesting.cs:11]: Is this the same as AdoNetInvariant?

File: test/Extensions/TesterAdoNet/RelationalUtilities/RelationalStorageForTesting.cs (hunk 1 above)
Comment (ElanHasson) [reply test/Extensions/TesterAdoNet/RelationalUtilities/RelationalStorageForTesting.cs:130]: So I tried to do that, then I got to that method which is static and used in a few places.

The static helper class, `TestContext`,  which exposes `WriteLine` was introduced via https://github.com/xunit/xunit/commit/850587bdbcbc0ea9e5a62a131787bd902332e747 which is xUnit and we are on 2.4.1.

image

If you're fine with it, I can pass `ITextOutputHelper` from the callers as a parameter, but that feels dirty. 

Is there anything stopping us from upgrading xunit? I imagine lots of breaking changes or you'd had done it already.

If upgrading is not an option, I can make our own `TestContext` :)

Comment (ReubenBond) [reply test/Extensions/TesterAdoNet/RelationalUtilities/RelationalStorageForTesting.cs:130]: Passing it as a parameter or deleting the line are both fine options to me

Comment (ReubenBond) [reply test/Extensions/TesterAdoNet/RelationalUtilities/RelationalStorageForTesting.cs:130]: Upgrading xUnit is fine, too

File: test/Extensions/TesterAdoNet/RelationalUtilities/PostgreSqlStorageForTesting.cs (hunk 2 above)
Comment (ElanHasson) [reply test/Extensions/TesterAdoNet/RelationalUtilities/PostgreSqlStorageForTesting.cs:11]: No, those are the full names of the drivers to use.

File: test/Extensions/TesterAdoNet/RelationalUtilities/RelationalStorageForTesting.cs (hunk 1 above)
Comment (ElanHasson) [reply test/Extensions/TesterAdoNet/RelationalUtilities/RelationalStorageForTesting.cs:130]: note: there are more than just that one `Console.WriteLine` in that method, sooo all or none. I'll figure it out.

Comment (ElanHasson) [reply test/Extensions/TesterAdoNet/RelationalUtilities/RelationalStorageForTesting.cs]: xunit.v3 isn't a thing yet apparently.

Yeah, that didn't go well. you can see the commit I made with the changes to remove `Console.WriteLine`. I'm not familiar enough with xUnit to know how to fix this. If you know, happy to re-revert and do it.

I just removed the offending line.

File: src/AdoNet/Orleans.Clustering.AdoNet/Messaging/AdoNetClusteringTable.cs [hunk 3]
```diff
@@ -181,31 +181,41 @@ public async Task DeleteMembershipTableEntries(string clusterId)
...
+            }
+            catch (Exception ex)
+            {
+                if (logger.IsEnabled(LogLevel.Trace)) logger.LogTrace(ex, "Insert silo membership version failed");
+                throw;
+            }
         }
     }
-}
+}
```
Comment (ElanHasson) [inline src/AdoNet/Orleans.Clustering.AdoNet/Messaging/AdoNetClusteringTable.cs:221]:
```suggestion
}

```

**Pull request merged**